Add documentation for HIP graphs (#3585)
ROCm-Developer-Tools/HIP: docs/how-to/hipgraph.rst

Graphs can also be created directly using the HIP graph API, giving more
fine-grained control over the graph. In this case, the graph nodes are created
explicitly, together with their parameters and dependencies, which specify the
edges of the graph, thereby forming the graph structure.

The nodes are represented by the generic ``hipGraphNode_t`` type. The actual
node type is implicitly defined by the specific function used to add the node to
the graph, for example ``hipGraphAddKernelNode``. See the
:doc:`HIP graph API documentation<../doxygen/html/group___graph>` for the
available functions, they are of type ``hipGraphAdd{Type}Node``. Each type of
node also has a predefined set of parameters depending on the operation, for
example ``hipKernelNodeParams`` for a kernel launch. See the
:doc:`documentation for the general hipGraphNodeParams type<../doxygen/html/structhip_graph_node_params>`
for a list of available parameter types and their members.

The general flow for explicitly creating a graph is usually:

#. Create a graph ``hipGraph_t``
#. Create the nodes and their parameters and adding them to the graph

>   #. Define a ``hipGraphNode_t``

Reviewer: Add new line here.

  #. Define the parameter struct for the desired operation, by explicitly setting the appropriate struct's members.
  #. Use the appropriate ``hipGraphAdd{Type}Node`` function to add the node to the graph.

>     #. The dependencies can be defined when adding the node to the graph, or afterwards by using ``hipGraphAddDependencies``

Reviewer: Add new line here.

performance is usually better when the memory is preallocated, so that it
does not need to be reallocated in every iteration.

The same rules as for normal memory allocations apply for memory allocated and
freed by nodes, meaning that the nodes that access memory allocated in a graph
must be ordered after allocation and before freeing.

Unlike the normal memory management API, which is controlled by host-side execution,
this enables the driver to take care of memory reuse and optimizations.
The lifetime of memory allocated in a graph begins when the execution reaches the
node allocating the memory, and ends when either reaching the corresponding
free node within the graph, or after graph execution when a corresponding
:cpp:func:`hipFreeAsync` or :cpp:func:`hipFree` call is reached.
The memory can also be freed with a free node in a different graph that is
associated with the same memory address.

These memory allocations can also be set up to allow access from multiple GPUs,
just like normal allocations. HIP then takes care of allocating and mapping the
memory to the GPUs. When capturing a graph from a stream, the node sets the

> accessibility according to ``hipMemPoolSetAccess`` at the time of capturing.

Reviewer: Use cpp:func here too? Referemce reorganize will make this work.

    }

Explicit graph creation
================================================================================

Graphs can also be created directly using the HIP graph API, giving more
fine-grained control over the graph. In this case, the graph nodes are created
explicitly, together with their parameters and dependencies, which specify the
edges of the graph, thereby forming the graph structure.

The nodes are represented by the generic ``hipGraphNode_t`` type. The actual
node type is implicitly defined by the specific function used to add the node to
the graph, for example ``hipGraphAddKernelNode``. See the
:doc:`HIP graph API documentation<../doxygen/html/group___graph>` for the
available functions, they are of type ``hipGraphAdd{Type}Node``. Each type of
node also has a predefined set of parameters depending on the operation, for
example ``hipKernelNodeParams`` for a kernel launch. See the
:doc:`documentation for the general hipGraphNodeParams type<../doxygen/html/structhip_graph_node_params>`
for a list of available parameter types and their members.

The general flow for explicitly creating a graph is usually:

#. Create a graph ``hipGraph_t``

> #. Create the nodes and their parameters and add them to the graph

Reviewer: Add new line here.

  #. Define a ``hipGraphNode_t``

>   #. Define the parameter struct for the desired operation, by explicitly setting the appropriate struct's members.

Reviewer: Add new line here.

>   #. Use the appropriate ``hipGraphAdd{Type}Node`` function to add the node to the graph.

Reviewer: Add new line here.

#. Create a :cpp:type:`hipGraphExec_t` executable instance of the graph template using :cpp:func:`hipGraphInstantiate`
#. Use :cpp:func:`hipGraphLaunch` to launch the executable graph to a stream
#. After execution completes free and destroy graph resources

The first two steps are the initial setup and only need to be executed once. First
step is the definition of the operations (nodes) and the dependencies (edges)
between them. The second step is the instantiation of the graph. This takes care
of validating and initializing the graph, to reduce the overhead when executing
the graph. The third step is the execution of the graph, which takes care of
launching all the kernels and executing the operations while respecting their
dependencies and necessary synchronizations as specified.

Because HIP graphs require some setup and initialization overhead before their
first execution, graphs only provide a benefit for workloads that require
many iterations to complete.

In both methods the :cpp:type:`hipGraph_t` template for a graph is used to define the graph.
In order to actually launch a graph, the template needs to be instantiated using

> :cpp:func:`hipGraphInstantiate`, which results in an actually executable graph of type :cpp:type:`hipGraphExec_t`.

Reviewer 2: ```suggestion
:cpp:func:`hipGraphInstantiate`, which results in an executable graph of type :cpp:type:`hipGraphExec_t`.
```

memory on the device or copying memory between the host and the device.
Whether you want to pre-allocate the memory or manage it within the graph
depends on the use-case. If the graph is executed in a tight loop the
performance is usually better when the memory is preallocated, so that it
does not need to be reallocated in every iteration.

The same rules as for normal memory allocations apply for memory allocated and
freed by nodes, meaning that the nodes that access memory allocated in a graph
must be ordered after allocation and before freeing.

Memory management within the graph enables the driver to take care of memory reuse and optimizations.
The lifetime of memory managed in a graph begins when the execution reaches the
node allocating the memory, and ends when either reaching the corresponding
free node within the graph, or after graph execution when a corresponding
:cpp:func:`hipFreeAsync` or :cpp:func:`hipFree` call is reached.
The memory can also be freed with a free node in a different graph that is
associated with the same memory address.

Unlike device memory that is not associated with a graph, this does not necessarily

> mean, that the freed memory is returned back to the operating system immediately.

Reviewer 2: ```suggestion
mean that the freed memory is returned back to the operating system immediately.
```

Graphs can retain a memory pool for quickly reusing memory within the graph.

> This can be especially useful, when memory is freed and reallocated later on

Reviewer 2: ```suggestion
This can be especially useful when memory is freed and reallocated later on
```

just like normal allocations. HIP then takes care of allocating and mapping the
memory to the GPUs. When capturing a graph from a stream, the node sets the
accessibility according to :cpp:func:`hipMemPoolSetAccess` at the time of capturing.

Capture graphs from a stream
================================================================================

The easy way to integrate HIP graphs into already existing code is to use
:cpp:func:`hipStreamBeginCapture` and :cpp:func:`hipStreamEndCapture` to obtain a :cpp:type:`hipGraph_t`
graph template that includes the captured operations.

When starting to capture operations for a graph using :cpp:func:`hipStreamBeginCapture`,
the operations assigned to the stream are captured into a graph instead of being
executed. The associated graph is returned when calling :cpp:func:`hipStreamEndCapture`, which
also stops capturing operations.
In order to capture to an already existing graph use :cpp:func:`hipStreamBeginCaptureToGraph`.

The functions assigned to the capturing stream are not executed, but instead are
captured and defined as nodes in the graph, to be run when the instantiated
graph is launched.

> Functions have to be associated with a stream in order to be captured.

Reviewer 2: ```suggestion
Functions must be associated with a stream in order to be captured in the graph.
```

> This means, that non-HIP API functions are not captured by default, but are

Reviewer 2: ```suggestion
This means that non-HIP API functions are not captured by default, but are
```

> executed as encountered, like without stream capture.

Reviewer 2: ```suggestion
executed as encountered as standard functions, rather than as part of the graph.
```

In order to assign host functions to a stream use
:cpp:func:`hipLaunchHostFunc`, as shown in the following code example.
They will then be captured and defined as a host node in the resulting graph,
and won't be executed when encountered.

Synchronous HIP API calls that are implicitly assigned to the default stream are

> not permitted while capturing a stream , and will return an error. This is

Reviewer 2: ```suggestion
not permitted while capturing a stream, and will return an error. This is
```

because they implicitly synchronize and cause a dependency that can not be
captured within the stream. This includes functions like :cpp:func:`hipMalloc`,
:cpp:func:`hipMemcpy` and :cpp:func:`hipFree`. In order to capture these to the stream, replace
them with the corresponding asynchronous calls like :cpp:func:`hipMallocAsync`, :cpp:func:`hipMemcpyAsync` or :cpp:func:`hipFreeAsync`.

> The general flow for using stream capture in order to get a graph template is:

Reviewer 2: ```suggestion
The general flow for using stream capture to create a graph template is:
```
